### PR TITLE
Extract projections in registers - Index

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Follow-up PR to https://github.com/arangodb/arangodb/pull/20023

Extract document attribute projections of `INDEX` nodes and `JOIN` nodes into individual output registers, instead of storing them in a combined VelocyPack object.
Previously, the projected attributes of a document were all stored together in a newly-built VelocyPack object. For example, if the projections "a" and "b" were used, a new VelocyPack object {"a":..., "b": ... } was built and stored in the ExecutionNode's main output register.

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [x] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.11: -
  - [ ] Backport for 3.10: -

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/Aql/JoinExecutor.cpp [hunk 1]
```diff
@@ -265,13 +305,28 @@ auto JoinExecutor::produceRows(AqlItemBlockInputRange& inputRange,
...
                                   _currentRow, docPtr);
           } else {
-            _projectionsBuilder.clear();
-            _projectionsBuilder.openObject(true);
-            idx.projections.toVelocyPackFromDocument(_projectionsBuilder, doc,
-                                                     &_trx);
-            _projectionsBuilder.close();
-            output.moveValueInto(_infos.indexes[k].documentOutputRegister,
-                                 _currentRow, _projectionsBuilder.slice());
+            if (!idx.hasProjectionsForRegisters) {
```
Comment (MBkkt) [inline arangod/Aql/JoinExecutor.cpp:308]: else if?

**Review by MBkkt — APPROVED** [review]

File: arangod/Aql/JoinNode.cpp [hunk 2]
```diff
@@ -415,6 +478,19 @@ std::vector<JoinNode::IndexInfo>& JoinNode::getIndexInfos() {
   return _indexInfos;
 }
 
+bool JoinNode::isDeterministic() {
+  for (auto const& it : _indexInfos) {
```
Comment (mchacki) [inline arangod/Aql/JoinNode.cpp:482]: Isn't this just equal to return false?

**Review by mchacki — COMMENTED** [review]
In general LGTM!

There is only one function which looks rather complicated but can only return false on all paths.
Is this intentional or should one exit actually be true?

File: arangod/Aql/JoinNode.cpp [hunk 3]
```diff
@@ -415,6 +478,19 @@ std::vector<JoinNode::IndexInfo>& JoinNode::getIndexInfos() {
...
+    if (it.condition != nullptr && it.condition->root() != nullptr &&
+        !it.condition->root()->isDeterministic()) {
+      return false;
+    }
+    if (it.filter != nullptr && !it.filter->isDeterministic()) {
+      return false;
+    }
+  }
+  return false;
```
Comment (jsteemann) [inline arangod/Aql/JoinNode.cpp:491]:
```suggestion
  return true;
```

File: arangod/Aql/JoinNode.cpp (hunk 2 above)
Comment (jsteemann) [reply arangod/Aql/JoinNode.cpp:482]: Indeed. Fixed!

**Review by mchacki — APPROVED** [review]
LGTM!

**Pull request merged**